% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/utilities3.R
\name{maf2cat3}
\alias{maf2cat3}
\title{Construct catalog matix from MAF file (without of Ref_Tri column)}
\usage{
maf2cat3(
  maf,
  ref.genome,
  fix.chr = TRUE,
  remove.duplicates = TRUE,
  progress.bar = TRUE
)
}
\arguments{
\item{maf}{MAF file or data frame of chromosome, position, ref, and alt alleles}

\item{ref.genome}{Reference genome of class \code{BSgenome}}

\item{fix.chr}{Modify chromosome names into \code{chr1, chr2, ...} for \code{ref.genome}}

\item{remove.duplicates}{Remove duplicate entries}

\item{progress.bar}{Display progress bar}
}
\value{
Catalog matrix
}
\description{
Construct catalog matix from MAF file (without of Ref_Tri column)
}
\examples{
library(BSgenome.Hsapiens.UCSC.hg19)
maf <- system.file('extdata', 'brca.maf', package = 'DeepSig')
x <- maf2cat3(maf, BSgenome.Hsapiens.UCSC.hg19)
head(x)

}
